flocker/node/agents/test/test_blockdevice.py

```diff
@@ -1177,144 +1197,124 @@ def _infer_volumes_for_test(node_uuid, node_hostname, compute_instance_id,
     return volumes
 
 
-class DiscoverVolumesMethod(Names):
-    """
-    Constants for determining volumes in calculate_changes tests.
-
-    :ivar GET_VOLUMES_FROM_API: Simply call ``list_volumes`` on the passed in
-        API.
-    :ivar INFER_VOLUMES_FROM_STATE: Call ``_infer_volumes_for_test`` to create
-        a list of volumes from the manually constructed cluster_state.
-    """
-    GET_VOLUMES_FROM_API = NamedConstant()
-    INFER_VOLUMES_FROM_STATE = NamedConstant()
-
-
 @implementer(ILocalState)
-class FakeBlockDeviceDeployerLocalState(PClass):
-    """
-    This class is a terrible idea and should be thrown out ASAP to be
-    replaced by the real thing.
+def local_state_from_shared_state(
+    node_state,
+    nonmanifest_datasets,
+    volumes,
+):
     """
-    node_state = field(type=NodeState, mandatory=True)
-    nonmanifest_datasets = field(type=NonManifestDatasets, mandatory=True)
-    volumes = pvector_field(BlockDeviceVolume)
-    datasets = pmap_field(UUID, DiscoveredDataset)
+    Convert the state as reported by
+    ``BlockDeviceDeployerLocalState.shared_state_changes`` to a
+    ``BlockDeviceDeployerLocalState`` instance.
+
+    This exists so that tests that provide the former can be used
+    with an implementation that expects the former.
+
+    .. warning:: This function is a terrible idea and should be thrown out ASAP
+    to be replaced by the real thing.
+    """
+    datasets = {}
+    for dataset_id, dataset in nonmanifest_datasets.items():
+        dataset_id = UUID(dataset_id)
+        if dataset_id in node_state.devices:
+            datasets[dataset_id] = DiscoveredDataset(
+                state=DatasetStates.ATTACHED,
+                dataset_id=dataset_id,
+                maximum_size=dataset.maximum_size or 0,
+                device_path=node_state.devices[dataset_id],
+                blockdevice_id=_create_blockdevice_id_for_test(dataset_id),
+            )
+        else:
+            datasets[dataset_id] = DiscoveredDataset(
+                state=DatasetStates.NON_MANIFEST,
+                dataset_id=dataset_id,
+                maximum_size=dataset.maximum_size or 0,
+                blockdevice_id=_create_blockdevice_id_for_test(dataset_id),
+            )
 
-    def shared_state_changes(self):
-        return self.node_state, self.nonmanifest_datasets
+    if node_state.manifestations is not None:
+        for dataset_id, manifestation in node_state.manifestations.items():
+            dataset_id = UUID(dataset_id)
+            datasets[dataset_id] = DiscoveredDataset(
+                state=DatasetStates.MOUNTED,
+                dataset_id=dataset_id,
+                maximum_size=manifestation.dataset.maximum_size or 0,
+                device_path=node_state.devices[dataset_id],
+                blockdevice_id=_create_blockdevice_id_for_test(dataset_id),
+                mount_point=node_state.paths[unicode(dataset_id)],
+            )
 
+    return BlockDeviceDeployerLocalState(
```

Comment:
Perhaps as a sanity check we could have an `assert` here that the resulting object's `shared_state_changes()` match the input ones?